streams/src/main/java/org/apache/kafka/streams/processor/internals/RecordCollectorImpl.java

```diff
@@ -57,7 +57,7 @@
 
     private final static String LOG_MESSAGE = "Error sending record (key {} value {} timestamp {}) to topic {} due to {}; " +
         "No more records will be sent and no more offsets will be recorded for this task.";
-    private final static String EXCEPTION_MESSAGE = "%sAbort sending since %s with a previous record (key %s value %s timestamp %d) to topic %s due to %s";
+    private final static String EXCEPTION_MESSAGE = "%sAbort sending to topic %s due to %s since %s with previous record having (timestamp %d key %s value %s)";
```

Comment:
The separation between the format string and the arguments makes it pretty hard to visually tell whether this is going to substitute the right things in the right places. Personally, I would rather inline the string, but not everyone would agree.

At least we can and should add unit tests to verify the right messages get logged in response to the right conditions. This wasn't always possible, but now we have the `org.apache.kafka.streams.processor.internals.testutil.LogCaptureAppender`. You can search the code for usages of that class to see how to write such tests.